### PR TITLE
Modify bin/apisix to support the SO_REUSEPORT

### LINKED ISSUE #342: CLI: check linux kernel and enable `REUSEPORT` if possible.
http://nginx.org/en/docs/http/ngx_http_core_module.html#listen

This option helps to evenly distribute requests to different workers.

### PR DESCRIPTION
**Pull request by Miss-you**
NOTE: Please read the Contributing.md guidelines before submitting your patch:

https://github.com/apache/incubator-apisix/blob/master/Contributing.md#how-to-add-a-new-feature-or-change-an-existing-one

### Summary

Modify bin/apisix to increase the capacity so that the generated nginx.conf file can open the reuseport configuration
    Judgment logic:
    1. If it is a mac system, do not open reuseport
    2. If it is a linux system and the kernel version is higher than 3.9.0, then turn on the reuserport switch, and the reuseport option will be added when generating nginx.conf> configuration

### Full changelog

bin/apisix

### Issues resolved

Fix #342

### CODE REVIEW

File: bin/apisix [hunk 1]
```diff
@@ -128,13 +128,26 @@ stream {
...
+        listen {*port*} reuseport;
+        {% end %}
+        {% for _, port in ipairs(stream_proxy.udp or {}) do %}
+        listen {*port*} udp reuseport;
+        {% end %}
+
+        {% else %} {% --if enable_reuseport %}
+
         {% for _, port in ipairs(stream_proxy.tcp or {}) do %}
         listen {*port*};
```
Comment (membphis) [inline bin/apisix]: `listen {*port*}  {% if enable_reuseport then %} reuseport {% end %};`

This style is simpler.

**Comment by Miss-you** [comment]
@moonming

**Comment by membphis** [comment]
@Miss-you https://github.com/apache/incubator-apisix/pull/1063 had been merged already.

**Comment by spacewander** [comment]
Strictly speaking, detect if a feature is available or not via kernel version is not reliable. Linux distributions may backport kernel patch to an old release (search for "SO_REUSEPORT backport for linux distribution"), not even mentioning self-patched kernel.

I think a better solution is to allow the users to configure it, since they know better about their servers than us. Maybe we can enable this feature by default, because most of the servers have kernel newer than 3.9.

**Comment by Miss-you** [comment]
Considering that the kernel version of the current production environment basically supports reuseport, I think it is more reasonable to open the REUSEPORT configuration file by default and supporting configuration it.

**Comment by Miss-you** [comment]
I remember that there are two main pits in reuseport, one is the kernel version, and the other is the support of http2 and http3.

**Comment by membphis** [comment]
> open the REUSEPORT configuration file by default and supporting configuration it.

this way is simpler. thx @spacewander

@Miss-you Expect your new commit ^_^

**Comment by Miss-you** [comment]
@membphis @spacewander 
hi, I have add the 'enable_reuseport' config into config.yaml. Please take a look.

File: bin/apisix [hunk 2]
```diff
@@ -507,6 +507,7 @@ local function init()
         with_module_status = with_module_status,
         node_ssl_listen = 9443,     -- default value
         error_log = {level = "warn"},
+        enable_reuseport = true, -- default true
```
Comment (moonming) [inline bin/apisix]: the default value should be set in `config.yaml`, not here.

File: bin/apisix [hunk 3]
```diff
@@ -530,6 +531,10 @@ local function init()
         sys_conf["worker_processes"] = "auto"
     end
 
+    if sys_conf["enable_reuseport"] == false then
+        enable_reuseport = false
+    end
+
```
Comment (moonming) [inline bin/apisix]: we can use the `enable_reuseport` in `config.yaml`

Comment (Miss-you) [reply bin/apisix]: done

File: bin/apisix (hunk 2 above)
Comment (Miss-you) [reply bin/apisix]: done

**Comment by membphis** [comment]
waiting your test case ^_^

File: conf/config.yaml [hunk 4]
```diff
@@ -21,7 +21,7 @@ apisix:
   enable_admin_cors: true         # Admin API support CORS response headers.
   enable_debug: false
   enable_dev_mode: false          # Sets nginx worker_processes to 1 if set to true
-  enable_reuseport: true          # Enable nginx SO_REUSEPORT switch if set to true.
+  enable_reuseport: false          # Enable nginx SO_REUSEPORT switch if set to true.
```
Comment (membphis) [inline conf/config.yaml]: default value `true` is better

Comment (Miss-you) [reply conf/config.yaml]: setting enable_reuseport false to verify that test cases are working.

Comment (membphis) [reply conf/config.yaml]: For testing, you can write a separate script to modify `config.yaml`.

**Comment by membphis** [comment]
image

**Comment by membphis** [comment]
ok, we can merge it to `master` branch first for checking it can work fine.

**Pull request merged**